Add DelayedFlipper and delayed options to EmptyRecyclerFlipper + ProgressEmptyRecyclerFlipper

bmoliveira opened the pull request:
This PR adds the option of adding the `DelayedFlipper` to `EmptyRecyclerFlipper` and `ProgressEmptyRecyclerFlipper`.

Additional changes:

- Migrate all Flippers to Kotlin.
- Refactor `EmptyRecyclerFlipper` and `ProgressEmptyRecyclerFlipper` to be configurable.
- Convert the project configuration into Kotlin DSL.

bmoliveira commented on Flippers/src/main/java/io/doist/recyclerviewext/flippers/EmptyRecyclerFlipper.kt:
```diff
@@ -0,0 +1,169 @@
+    private val adapterObserver: AdapterDataObserver = EmptyAdapterDataObserver()
+    private var currentItemCount = 0
+
+    @Deprecated("Flipper access will be removed, customize flipper with configure method.")
```
I've added to have some compatibility with the old class being a direct descendant of Flipper and now it does not have the Flipper methods. Do you have any better recommendation to implement this?

JinneeJ replied:
I'd recommend to have `DelayedEmptyRecyclerFlipper` and keep API of `EmptyRecyclerFlipper` unchanged :)

bmoliveira replied:
I'll add the Delayed versions as a separate class, but I'm maintaining the configure approach, so the flipper can be configured with a lifecycle so it can clean itself up automatically

JinneeJ commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,6 @@
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.6-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.5-bin.zip
```
Why would we update it only to `6.5` given that latest version is `6.8.3`? Also I'd recommend to keep `all` to have sources available.

JinneeJ commented on settings.gradle.kts:
```diff
@@ -0,0 +1,11 @@
+include (
+        ":Animations",
+        ":Dividers",
+        ":DragDrop",
+        ":ClickListeners",
+        ":Flippers",
+        ":Selectors",
+        ":StickyHeaders",
+        ":PinchZoom"
```
I think indent is bigger that required by code styles 🙃

JinneeJ reviewed (changes requested):
Thanks for adding this! I've left some comments in the code but I didn't review thoroughly since I'd like to discuss some general things regarding changes:

1. I think that we shouldn't convert to Kotlin here. Reasons:
    - Users of this library might use gradle version that doesn't properly support kts.
    - Given that users of this library might not use Kotlin at all, we shouldn't add kotlin stdlibs and coroutines dependencies.
2. Did you consider to not use coroutines at all? Can we achieve the same effect with handler?
3. Do we need to introduce `DelayedFlipper` at all? I'm not sure what is the main use case for it but may be it can be achieved with combination of `ProgressEmptyRecyclerFlipper` with `ContentloadingProgressBar`? :)

bmoliveira commented on Flippers/build.gradle.kts:
```diff
@@ -0,0 +1,46 @@
+
+dependencies {
+    implementation("androidx.recyclerview:recyclerview:1.1.0")
+    api("androidx.appcompat:appcompat:1.2.0")
+    api("org.jetbrains.kotlinx:kotlinx-coroutines-android:1.4.2")
+    api("androidx.lifecycle:lifecycle-runtime-ktx:2.3.0")
+    api("androidx.core:core-ktx:1.3.2")
```
I'm removing all the unnecessary libs. The lifecycle will be important because I'm implementing lifecycle observing in order to not retain things after the view is destroyed.

bmoliveira commented on Flippers/src/main/java/io/doist/recyclerviewext/flippers/DefaultFlipperAnimator.kt:
```diff
@@ -0,0 +1,49 @@
+import android.animation.Animator
+import android.animation.AnimatorListenerAdapter
+import android.view.View
+import androidx.core.view.isVisible
```
Yeah it is a giant overkill for this 😅

bmoliveira commented:
I will revert the `.kts` conversion, it makes sense to support the wider range of environments as possible.

Regarding Kotlin usage at this stage, it should not be a problem since almost every project will end up using it. And the versatility and benefits it brings are greater than the downsides.

JinneeJ commented on Flippers/build.gradle:
```diff
@@ -1,17 +1,18 @@
 dependencies {
-    implementation "androidx.recyclerview:recyclerview:1.0.0"
+    implementation "androidx.recyclerview:recyclerview:1.1.0"
+    api "androidx.lifecycle:lifecycle-runtime-ktx:2.3.0"
```
I think we shouldn't have `api` here and it should be used only in private modules or in such cases like ktx modules, which are in fact wrappers around java modules and it's expected to use `api` there 😊

JinneeJ commented on gradle.properties:
```diff
@@ -1,18 +1,21 @@
\ No newline at end of file
+# org.gradle.parallel=true
+#Thu Mar 25 17:19:38 WET 2021
+org.gradle.parallel=true
+org.gradle.configureondemand=true
+org.gradle.caching=true
+org.gradle.jvmargs=-Xmx3072m -XX:MaxPermSize=768m
+org.gradle.vfs.watch=true
```
I think we shouldn't enforce gradle configuration for developers but we can add recommendation and have comment about this here :)

JinneeJ commented on gradle.properties:
```diff
@@ -1,18 +1,21 @@
-# org.gradle.parallel=true
\ No newline at end of file
+# org.gradle.parallel=true
+#Thu Mar 25 17:19:38 WET 2021
```
Why do we need date here? 🙃

JinneeJ commented on Flippers/src/main/java/io/doist/recyclerviewext/flippers/ViewExt.kt:
```diff
@@ -0,0 +1,9 @@
+
+import android.view.View
+
+inline var View.isVisible: Boolean
+    get() = visibility == View.VISIBLE
+    set(value) {
+        visibility = if (value) View.VISIBLE else View.GONE
+    }
```
It will lead that IDE shows multiple options to import `View.isVisible` for projects that have both `RecyclerViewExtension` and `core-ctx` :) May be we can remove it or make it internal at least?

bmoliveira replied:
Made it internal, nice catch! 👌

JinneeJ commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.6-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.8.3-all.zip
```
I think you forgot to run wrapper task :)

JinneeJ reviewed (changes requested):
Thanks for changes! 👏

I still see two blocking things for me here:

1. I still think that it's still too early for Kotlin conversion. As far as I know among most popular Android libraries, only `okhttp` was converted to Kotlin. For example `AndroidX` is still in java and has Kotlin only in `ktx` modules. Conversion into Kotlin won't provide any advantages to lib users but only extra dependencies. Given that we don't maintain this library often, keeping it in Java doesn't slow down us.
2. I still don't understand why we need method `configure`. Given how much changes it brings to client's code, I'd avoid it. We can pass `livecycleOwner` into constructor along with views. If method `configure` is necessary for delayed version of flippers, then we can make this method protected and use it internally in flippers where needed.

In result to address both blocking problems I suggest following:
- Have separate module, e.g.`FlippersDelayed` 
- This module would depend on module `Flippers`
- Module `FlippersDelayed` would use Kotlin and lifecycle
- Keep API of module `Flippers` unchanged

It will allow to have modern implementation with Kotlin and lifecycle for delayed flippers and keep old API unchanged and without extra dependencies.

bmoliveira commented:
> I still think that it's still too early for Kotlin conversion. As far as I know among most popular Android libraries, only okhttp was converted to Kotlin. For example AndroidX is still in java and has Kotlin only in ktx modules. Conversion into Kotlin won't provide any advantages to lib users but only extra dependencies. Given that we don't maintain this library often, keeping it in Java doesn't slow down us.

You are right we don't need to force the users to Kotlin yet, and the maintenance here is marginal so I've refactored the `Flipper` classes back into Java.

> I still don't understand why we need method configure. Given how much changes it brings to client's code, I'd avoid it. We can pass livecycleOwner into constructor along with views. If method configure is necessary for delayed version of flippers, then we can make this method protected and use it internally in flippers where needed.

Here I was trying to refactor the `EmptyRecyclerFlipper` and `ProgressEmptyRecyclerFlipper` into a delegate approach because in some cases it might be needed, but the changes as you said are many, so I've maintained the current constructor approach and on the `Delayed` versions maintained the API as well.


> Have separate module, e.g.FlippersDelayed

I've refactored the modules into a separate one, but now I have a small problem that I didn't find a solution for. The `FlippersDelayed` module will depend on `Flippers`. Afterward, when we will use it if the project doesn't have a `Flippers` module it will not be able to resolve. With libraries being distributed via `maven` for example, we don't need to worry about this, but in the `git submodules` case, I can't find a workaround for it. [ref](https://github.com/Doist/Twist-Android/pull/523#discussion_r604900236)

JinneeJ commented on Flippers/src/main/java/io/doist/recyclerviewext/flippers/EmptyRecyclerFlipper.java:
```diff
@@ -9,29 +9,46 @@
+ * ListActivity} for {@link ListView}.
  */
-public class EmptyRecyclerFlipper extends Flipper {
+public class EmptyRecyclerFlipper {
```
Now it doesn't extend `Flipper` anymore. Can we avoid it? Otherwise client's code can become broken it it casts `EmptyRecyclerFlipper` to `Flipper` :)

bmoliveira replied:
I was trying to avoid subclassing and refactored to composing. But it would break clients code, so I've added back the subclassing and added warnings of deprecation on the methods and on the README :)

JinneeJ replied:
> I was trying to avoid subclassing in favor of composing.

I think choice subclassing vs composing was made when this library was implemented. I think it's late to change this 😅

JinneeJ reviewed (changes requested):
Thanks a lot for changes! 🙏 

Found one another issue that might break clients code :)

> I've refactored the modules into a separate one, but now I have a small problem that I didn't find a solution for. The FlippersDelayed module will depend on Flippers. Afterward, when we will use it if the project doesn't have a Flippers module it will not be able to resolve. With libraries being distributed via maven for example, we don't need to worry about this, but in the git submodules case, I can't find a workaround for it.

I think it's not a big issue. To avoid any misunderstanding we can mark dependency as `implementation` and left a note in readme for `FlippersDelayed` that dependency on `Flippers` should be specified explicitly. Having said this, I think that if we continue to support this library, we should consider publishing it into maven :)

bmoliveira commented:
> we should consider publishing it into maven :)

I also agree it would make the versioning and release process clearer for the end users.

JinneeJ commented on FlippersDelayed/src/main/java/io/doist/recyclerviewext/flippers/DelayedProgressEmptyRecyclerFlipper.kt:
```diff
@@ -0,0 +1,127 @@
+            else -> {
+                delayedFlipper.replaceNoAnimation(
+                    outView, inView
+                ) { onLoadingFlipped(outView, inView, false) }
```
Let's use the same formatting as in previous clauses :)

JinneeJ reviewed (approved):
Thanks for changes! 👏 Looks good, left some nitpicking :)

JinneeJ commented on Flippers/src/main/java/io/doist/recyclerviewext/flippers/EmptyRecyclerFlipper.java:
```diff
@@ -45,36 +67,129 @@ public boolean monitor(RecyclerView.Adapter adapter) {
+    }
+
+    /**
+     * @deprecated To customize the flipper use the {@link #getFlipper()}.
```
Why we deprecate methods `replace*`? They are not about cusomization :)

bmoliveira replied:
I was deprecating these methods so we could remove the subclassing from `Flipper` but as you said:

> I think choice subclassing vs composing was made when this library was implemented. I think it's late to change this 😅

I've just let them as an interface for the inner `Flipper` without warning.

JinneeJ reviewed (changes requested):
Sorry, I missed some unclear deprecation during previous review :)

bmoliveira commented on FlippersDelayed/src/main/java/io/doist/recyclerviewext/flippers/DelayedEmptyRecyclerFlipper.kt:
```diff
@@ -0,0 +1,66 @@
+
+    init {
+        lifecycleOwner.lifecycle.addObserver(lifecycleObserver)
+    }
```
No, I've overlapped the responsibilities of both classes. I've removed the lifecycle observing from `EmptyRecyclerFlipper` and `ProgressEmptuRecyclerFlipper` since it was only listening to what `DelayedFlipper` does.

JinneeJ reviewed (approved):
Thanks for changes! 🎉
